Adding the ability to pick the label applied to an automatic reference

martinthomson opened the pull request:
This changes the syntax for `{{?RFCXXXX}}` and `{{!RFCBLAH}}` to allow
the author to specify a name for the reference.  This is added after a
space, as in `{{?RFC5389 STUN}}`.  This affects the label used if symbolic
references are enabled, e.g., `[STUN]`.  This is especially handy for DOI
references, which I just added because it keeps the symbolic reference
tidy.  For example, `{{?DOI.10.1145/1282427.1282421 SST}}`, which is a
reference we are using in QUIC.

The first half of the reference only needs to be used once.  You can refer
to `{{? SST}}` many times as long as there is a single definition that
has the full reference.  This was close to free, and may serve to replace
the abbreviation support for many.

I have basically no testing on `<iref>`.

This allowed me to consolidate the places where you were testing for
labels that started with a digit.  That too was necessary to have a
single location where the DOI reference was mangled.

martinthomson commented on bin/kramdown-rfc2629:
```diff
@@ -71,21 +71,30 @@ def xml_from_sections(input)
   sechash.each do |k, v|
     next if k == "fluff"
-    v.gsub!(/{{(?:([?!])(-)?|(-))([\w.\-]+)}}/) do |match|
+    v.gsub!(/{{(?:([?!])(-)?|(-))([\w.\/_\-]+(?: ([\w._\-]+))?)}}/) do |match|
```
Testing shows that this works, but this won't match on `{{? SST}}`.  I wonder if this should be `/{{(?:([?!])(-)?|(-))((?:[\w.\/_\-]+)?(?: ([\w._\-]+))?)}}/` instead.

martinthomson replied:
Belay that comment, clearly this works as shown below.

martinthomson commented on bin/kramdown-rfc2629:
```diff
@@ -124,21 +137,22 @@ def xml_from_sections(input)
         else
           unless v && Hash === v
-            warn "*** don't know how to expand ref #{k}" 
+            warn "*** don't know how to expand ref #{k}"
```
Bah, trailing whitespace.

martinthomson commented:
@cabo, do you have any opinions on this feature?  I would like to be able to use DOIs for a reference in QUIC, but this is blocking it.  I could contribute a much smaller patch that just added DOI support, but this is much nicer.

cabo commented:
Yes, I do.  It's about time kramdown-rfc grows this feature!
There are a couple of things I think need to be done before we can mainline this.  This week I have a heavy week of teaching, so I hope I can provide feedback next week.

cabo commented:
Long week.

There are two things I'd like to obtain to be able to mainline this:

* This feature needs to be available from the YAML header, not just from the YAML header autogeneration ({{?...}})

* spaces in markup elements are icky as they turn into line breaks when reformatting source.

To address the latter, I would probably go for a syntax such as {{?XMPP=I-D.ietf-xmpp-3920bis}} -- this can then simply be referenced as {{XMPP}} in the rest of the document.

To address the former, I'd probably add an attribute "bib" or some such:

```
normative:
  XMPP:
    bib: I-D.ietf-xmpp-3920bis
```

that of course defaults to the anchor name.

(Now, that attribute could also provide a way to go beyond the current mapping in XML_RESOURCE_ORG_MAP.)

martinthomson commented:
Yeah, long week is an understatement.  Thanks for looking at this.

I've rebased and changed to your recommended syntax (I wasn't happy with the original, so that was a welcome suggestion).  I decided to use the same syntax for the YAML part as well.  That also makes it easier to tell if you have a real definition for the reference.  You can see from the examples what this looks like.

martinthomson commented:
Oh, one thing I realized after I finished up yesterday is that there is perhaps another alternative to this that is superior, though I won't have time to implement this.

I chose to do this:

```
informative:
  LOSS-PROBE=I-D.dukkipati-tcpm-tcp-loss-probe:
```

But this might also work:

```
informative:
  LOSS-PROBE: I-D.dukkipati-tcpm-tcp-loss-probe
```

The former is nice in that the syntax is the same throughout, but I had to jump through some hoops to get it to work cleanly.  The latter might be superior implementation-wise.  I can implement it if you would prefer that.

cabo commented:
That place is already taken for the local label:

```
RFC7252: coap
```

is a shortcut for

```
RFC7252:
  -: coap
```

which is the thing that lets you reference RFC7252 as `{{-coap}}` -- a nice way if you don't want to change internal references each time a document transitions from individual to WG document to RFC, without exposing that internal name in the actual draft.

What I would like to add is the bib thing I had in the above example, what you called href in your code.
(What you call label, I call anchor.)

Assuming it is OK to use a hash key of "=" for that (we could also use "bib" or "href"), this could become:

```
informative:
  LOSS-PROBE: 
    =: I-D.dukkipati-tcpm-tcp-loss-probe
```

(and you could even add a "-" to that as well, all though swapping out both anchor and local name doesn't seem to make much sense).

I have some more teaching to do tomorrow, but probably can throw all this together from your code over the weekend.

Other things i'd like to have:

* being able to be explicit about the bibxmlX (i.e., not infer that path name component, but write it down if that helps).  We would still cache everything in one directory. (.refcache or ~/.cache/xml2rfc).

maybe: 
```
    =: bibxml3/I-D.dukkipati-tcpm-tcp-loss-probe
```

* (Hmm, I forgot the other one.)

cabo commented:
Of course, I could hack it that if the "local name" contains a dot, it is really a bib name.  Hmm, does not work for RFC that well, but even that could be special-cased.  Maybe too much magic.

```
normative:
  TCP: RFC0793
  RFC7252: coap
  URIDESIGN:
    =: RFC7320
    -: lawn
...

{{TCP}} ... {{-coap}} ... {{-lawn}}
```

➔ [TCP].... [RFC7252] ... [URIDESIGN]

(Hmm.)

cabo commented on lib/kramdown-rfc2629.rb:
```diff
@@ -39,14 +39,20 @@ def initialize(*doc)
+      def self.split_xref(t)
+        label, href = t.split('=')
+        label = label.sub(/\A\d/) { "_#{$&}" } # can't start an IDREF with a number
+        label = label.sub('DOI.', 'DOI_').gsub('/', '_') # DOIs are gross
```
Why get rid of the dot?  It doesn't seem to be a problem with I-D.foo and other references.

martinthomson replied:
Yeah, it's unnecessary.

cabo commented on lib/kramdown-rfc2629.rb:
```diff
@@ -495,8 +501,10 @@ def convert_img(el, indent, opts) # misuse the tag!
+          warn "*** no label for anchor #{anchor}" unless label
           to_insert = ""
-          anchor.scan(/(W3C|3GPP|[A-Z-]+)[.]?([A-Za-z0-9.-]+)/) do |t, n|
+          anchor.scan(/(W3C|3GPP|[A-Z-]+)[.]?([A-Za-z0-9.\/-]+)/) do |t, n|
```
anchors are XML names, and / is not actually allowed in an XML name, so  DOI anchors will always need to have an _ for the /.  (This is not a problem if the DOI reference is used with a custom anchor, hmm.)

martinthomson replied:
Yeah, I note an example on xml2rfc.tools.ietf.org that uses `+`, but I didn't add that.  I'll let you do that part.

martinthomson replied:
...and I see that you have done so already as I rebase.  Excellent.

martinthomson replied:
Or not, added +

cabo commented on lib/kramdown-rfc2629.rb:
```diff
@@ -494,13 +500,18 @@ def convert_img(el, indent, opts) # misuse the tag!
+          end
           to_insert = ""
-          anchor.scan(/(W3C|3GPP|[A-Z-]+)[.]?([A-Za-z_0-9.-]+)/) do |t, n|
+          src.scan(/(W3C|3GPP|[A-Z-]+)[.]?([A-Za-z_0-9.\/\+-]+)/) do |t, n|
```
```
 play.xml: Line 102: Syntax of value for attribute target of xref is not valid
 play.xml: Line 796: Syntax of value for attribute anchor of reference is not valid
```

102: `interest both in the Web graphics community [TypedArray] = <xref target="Typed+Array"/> and in`

796: `<reference anchor="Typed+Array" target="https://www.khronos.org/registry/typedarray/specs/1.0/">`

So, a + sign cannot be part of an XML anchor, but of course it can be part of a reference URL.

More importantly, putting a '/' in the reference URL confuses the caching logic as only the base name is used for the filename.

martinthomson replied:
Yeah, I think that requires sanitizing the cache filename.  The anchor should be sanitized already.

cabo commented on lib/kramdown-rfc2629.rb:
```diff
@@ -56,8 +62,8 @@ def parse_xref
-        href = @src[1]
-        el = Element.new(:iref, nil, {'target' => href}) # XXX
+        anchor, = RFC2629Kramdown.split_xref(@src[1])
+        el = Element.new(:iref, nil, {'target' => anchor}) # XXX
```
Not sure why this needs to be done to index entries.

martinthomson replied:
Yep, you are right.  I was hoping to take advantage of the cleanup that is done in the split function, but it changes the entries in the index in a bad way.

cabo commented:
The one thing that stopped me in me track yesterday was that I don't know what to do for `stand_alone: false` (i.e., where entity references are generated).  Clearly, label picking cannot be supported with that setting.  For a while, I tried coming up with a way a label picking document at least compiles, even if the label picks are ignored.  But it seems, without invasive changes in alias handling, this cannot be done.  If that is so, so be it (in that case, we'll just need an error message).

martinthomson commented:
Yeah, `standalone: false` is a trap.  I've learned that giving people options is only good if they need those options.  I'm struggling to find a reason that `standalone: false` is worthwhile here, other than the fact that you can "build" offline.  Your new caching options are superior in that regard, of course.

cabo commented:
`stand_alone: false` is part of the "You can take your data elsewhere" promise of kramdown-rfc.
If you are ever fed up with authoring in markdown, just make this setting, process one last time, and you have a fully editable XML again that automatically updates draft versions from the XML.

So if there has been some label picking, I think it is better to generate a `stand_alone: false` XML that requires some manual touch-up than to give up.  So I would turn the error into a warning.

cabo commented on lib/kramdown-rfc2629.rb:
```diff
@@ -494,19 +499,24 @@ def convert_img(el, indent, opts) # misuse the tag!
             puts "Huh: ${fn}" unless sub
             url = "#{XML_RESOURCE_ORG_PREFIX}/#{sub}/#{fn}"
-            to_insert = get_and_cache_resource(url, ttl)
+            to_insert = get_and_cache_resource(url, fn.gsub('/', '_').gsub('+', '_'), ttl)
```
Why avoid "+" in cache file names?

martinthomson replied:
I've not had good luck with odd characters in filenames and it's one that causes issues in shells and URLs. I can revert it if you like.

cabo commented on bin/kramdown-rfc2629:
```diff
@@ -52,10 +52,12 @@ def xml_from_sections(input)
   [:ref, :normative, :informative].each do |sn|
     if refs = ps.has(sn)
       warn "*** bad section #{sn}: #{refs.inspect}" unless refs.respond_to? :each
+      aliased_refs = { }
```
unused

cabo commented:
I think I have cherry-picked all I needed from this PR in 1.1.0 -- please tell me if I missed any functionality.  Thank you!
